### PR TITLE
fix: ensure that each js file served up by vite dev server has an inline sourcemap

### PR DESCRIPTION
**Pull request by ryanthemanuel**
### Additional details

We have a custom vite dev server plugin that transforms JS files to include inlined source maps but there were two scenarios that were missed with this plugin:

- files that include a cache busting query parameter (e.g. ?v=12345)
- files that include a sourcemap but one that is not inline

This PR resolves both of these scenarios.

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Review by AtofStryker — COMMENTED** [review]
does this fix the code frame issue we were seeing with vite and component testing?

**Comment by ryanthemanuel** [comment]
> does this fix the code frame issue we were seeing with vite and component testing?

Unfortunately, no that is more complicated to resolve however this will get us where we need to get to in order to properly have sourcemaps available in the protocol space.

File: npm/vite-dev-server/src/plugins/sourcemap.ts [hunk 1]
```diff
@@ -15,7 +15,11 @@ export const CypressSourcemap = (
     enforce: 'post',
     transform (code, id, options?) {
       try {
-        if (/\.js$/i.test(id) && !/\/\/# sourceMappingURL=/i.test(code)) {
+        // Remove query parameters from the id. This is necessary because some files
+        // have a cache buster query parameter (e.g. `?v=12345`)
+        const queryParameterLessId = id.split('?')[0]
+
+        if (/\.js$/i.test(queryParameterLessId) && !/\/\/# sourceMappingURL=data/i.test(code)) {
```
Comment (mschile) [inline npm/vite-dev-server/src/plugins/sourcemap.ts]: Suggest adding a comment here stating that we are looking for files without a sourcemap or without a non-inline sourcemap.

File: npm/vite-dev-server/src/plugins/sourcemap.ts [hunk 2]
```diff
@@ -35,8 +39,13 @@ export const CypressSourcemap = (
           */
 
           const sourcemap = this.getCombinedSourcemap()
+          const sourcemapUrl = sourcemap.toUrl()
 
-          code += `\n//# sourceMappingURL=${sourcemap.toUrl()}`
+          if (/\/\/# sourceMappingURL=/i.test(code)) {
```
Comment (mschile) [inline npm/vite-dev-server/src/plugins/sourcemap.ts:37]: Suggest adding a comment here.

File: npm/vite-dev-server/test/plugins/sourcemap.spec.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import { Plugin } from 'vite-5'
+import { ViteDevServerConfig } from '../../src/devServer'
+import { Vite } from '../../src/getVite'
+import { CypressSourcemap } from '../../src/plugins'
+import Chai, { expect } from 'chai'
+import SinonChai from 'sinon-chai'
+
+Chai.use(SinonChai)
+
+describe('sourcemap plugin', () => {
```
Comment (mschile) [inline npm/vite-dev-server/test/plugins/sourcemap.spec.ts:11]: Suggest also adding a test to verify non-js files don't get the source map url appended and another test to verify a js file that already contains an inline source map doesn't get updated.

Comment (ryanthemanuel) [reply npm/vite-dev-server/test/plugins/sourcemap.spec.ts:11]: Good call: [`8607541` (#30606)](https://github.com/cypress-io/cypress/pull/30606/commits/86075410a545f84215dcaa369b0c83ce9b54a9df)

File: npm/vite-dev-server/src/plugins/sourcemap.ts (hunk 2 above)
Comment (ryanthemanuel) [reply npm/vite-dev-server/src/plugins/sourcemap.ts:37]: Comment added: [`e076083` (#30606)](https://github.com/cypress-io/cypress/pull/30606/commits/e07608342f1c130d602a90f8e9167753f835a5f0)

File: npm/vite-dev-server/src/plugins/sourcemap.ts (hunk 1 above)
Comment (ryanthemanuel) [reply npm/vite-dev-server/src/plugins/sourcemap.ts]: Added: [`91bf502` (#30606)](https://github.com/cypress-io/cypress/pull/30606/commits/91bf5021a1fcb18a7621f04e83a8129d4cdafcc8)

File: npm/vite-dev-server/src/plugins/sourcemap.ts [hunk 4]
```diff
@@ -15,28 +15,33 @@ export const CypressSourcemap = (
...
-          sourcemap mapping 1-to-1 to the output file. We can grab this and pass it along as a sourcemap
-          we want Vite to embed into the output, giving Cypress a sourcemap to use for codeFrame lookups.
-          @see https://rollupjs.org/guide/en/#thisgetcombinedsourcemap
+        // Remove query parameters from the id. This is necessary because some files
+        // have a cache buster query parameter (e.g. `?v=12345`)
+        const queryParameterLessId = id.split('?')[0]
 
-          We utilize a 'post' plugin here and manually append the sourcemap content to the code. We *should*
+        // Check if the file has a JavaScript extension and does not have an inlined sourcemap
+        if (/\.(js|jsx|ts|tsx|vue|mjs|cjs)$/i.test(queryParameterLessId) && !/\/\/# sourceMappingURL=data/i.test(code)) {
```
Comment (ryanthemanuel) [inline npm/vite-dev-server/src/plugins/sourcemap.ts]:
```suggestion
        if (/\.(js|jsx|ts|tsx|vue|svelte|mjs|cjs)$/i.test(queryParameterLessId) && !/\/\/# sourceMappingURL=data/i.test(code)) {
```

**Review by mschile — APPROVED** [review]

File: npm/vite-dev-server/test/plugins/sourcemap.spec.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+import { Vite } from '../../src/getVite'
+import { CypressSourcemap } from '../../src/plugins'
+import Chai, { expect } from 'chai'
+import SinonChai from 'sinon-chai'
+import sinon from 'sinon'
+
+Chai.use(SinonChai)
+
+describe('sourcemap plugin', () => {
+  ['js', 'jsx', 'ts', 'tsx', 'vue', 'mjs', 'cjs'].forEach((ext) => {
```
Comment (ryanthemanuel) [inline npm/vite-dev-server/test/plugins/sourcemap.spec.ts]:
```suggestion
  ['js', 'jsx', 'ts', 'tsx', 'vue', 'svelte', 'mjs', 'cjs'].forEach((ext) => {
```

**Review by AtofStryker — APPROVED** [review]

**Pull request merged**